tests/functional/test_new_resolver_hashes.py

```diff
@@ -0,0 +1,125 @@
+import collections
+import hashlib
+
+import pytest
+
+from pip._internal.utils.urls import path_to_url
+from tests.lib import (
+    create_basic_sdist_for_package,
+    create_basic_wheel_for_package,
+)
+
+_FindLinks = collections.namedtuple(
+    "_FindLinks", "index_html sdist_hash wheel_hash",
+)
+
+
+def _create_find_links(script):
+    sdist_path = create_basic_sdist_for_package(script, "base", "0.1.0")
+    wheel_path = create_basic_wheel_for_package(script, "base", "0.1.0")
+
+    sdist_hash = hashlib.sha256(sdist_path.read_bytes()).hexdigest()
+    wheel_hash = hashlib.sha256(wheel_path.read_bytes()).hexdigest()
+
+    index_html = script.scratch_path / "index.html"
+    index_html.write_text(
+        """
+        <a href="{sdist_url}#sha256={sdist_hash}">{sdist_path.stem}</a>
+        <a href="{wheel_url}#sha256={wheel_hash}">{wheel_path.stem}</a>
+        """.format(
+            sdist_url=path_to_url(sdist_path),
+            sdist_hash=sdist_hash,
+            sdist_path=sdist_path,
+            wheel_url=path_to_url(wheel_path),
+            wheel_hash=wheel_hash,
+            wheel_path=wheel_path,
+        )
+    )
+
+    return _FindLinks(index_html, sdist_hash, wheel_hash)
+
+
+@pytest.mark.parametrize(
+    "requirements_template, message",
+    [
+        (
+            """
+            base==0.1.0 --hash=sha256:{sdist_hash} --hash=sha256:{wheel_hash}
+            base==0.1.0 --hash=sha256:{sdist_hash} --hash=sha256:{wheel_hash}
+            """,
+            "Checked 2 links for project {name!r} against 2 hashes "
+            "(2 matches, 0 no digest): discarding no candidates",
+        ),
+        (
+            # Different hash lists are intersected.
+            """
+            base==0.1.0 --hash=sha256:{sdist_hash} --hash=sha256:{wheel_hash}
+            base==0.1.0 --hash=sha256:{sdist_hash}
+            """,
+            "Checked 2 links for project {name!r} against 1 hashes "
+            "(1 matches, 0 no digest): discarding 1 non-matches",
+        ),
+    ],
+    ids=["identical", "intersect"],
```

Comment:
Can we add a case of the intersection being empty, to ensure we test the error message as well?

Reply:
Do you mean a test for something like this?

```
base==0.1.0
base==0.1.0 --hash=... --hash=...
```

The error you get here is unrelated to the intersection logic. This input is rejected straight on parsing since the requirements parser checks every requirement line in the file contains at least one hash. This behaviour definitely needs to be tested (I’m not sure if it is currently), but not here IMO. It is a part of the requirements file parser, not the resolver.

Reply:
I meant something like

```
base == 1.0.0 --hash={sdist_hash}
base == 1.0.0 --hash={wheel_hash}
```

The intersection of these hashes would be empty, which is what I'm suggesting it might be worthwhile to test.

Reply:
I see, that makes sense. Will update soon.

Reply:
Hmm, so `pip install` fails as expected, but the error message seems suboptimal. Since the two lines are two separate requirements, the error users see is

```
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE.
If you have updated the package versions, please update the hashes. Otherwise,
examine the package contents carefully; someone may have tampered with them.
     base==0.1.0 from ...:
         Expected sha256 {wheel_hash}
              Got        {sdist_hash}
```

(The hashes in *Expected* and *Got* flip if you flip the requirement lines.)

But improving the error message seems out of scope of this fix (and I’m not even sure how to improve it). Maybe I should just test the errors as-is and open a new issue instead.